### PR TITLE
Distinguish all-float blocks and float arrays

### PR DESCRIPTION
**Pull request by mshinwell**
Flambda 2 has a strict distinction between blocks (including records) and arrays.  At first we tried to treat these together but that proved to be a bad idea.

One of the changes required to the frontend to enforce this distinction is given here.  It distinguishes all-float records from float arrays.  The change to `Lambda.structured_constant` necessitated a bootstrap.

### CODE REVIEW

File: middle_end/closure/closure.ml [hunk 1]
```diff
@@ -900,6 +900,9 @@ let rec close ({ backend; fenv; cenv ; mutable_vars } as env) lam =
         | Const_pointer n -> Uconst_ptr n
         | Const_block (tag, fields) ->
             str (Uconst_block (tag, List.map transl fields))
+        | Const_float_block sl ->
+            (* CR mshinwell: Add [Uconst_float_block]? *)
```
Comment (stedolan) [inline middle_end/closure/closure.ml:904]: I don't think there's any real need to do this. I don't think Clambda needs to distinguish the two, and if one day it needs to it's easily changed then.

File: middle_end/flambda/simplify_primitives.ml [hunk 2]
```diff
@@ -123,6 +123,7 @@ let primitive (p : Clambda_primitives.primitive) (args, approxs)
   | Pmakearray(_, _) when is_empty approxs ->
     Prim (Pmakeblock(0, Asttypes.Immutable, Some []), [], dbg),
     A.value_block (Tag.create_exn 0) [||], C.Benefit.zero
+  (* CR mshinwell: Work out what to do here with [Pmakefloatblock] *)
```
Comment (stedolan) [inline middle_end/flambda/simplify_primitives.ml:126]: Do you plan something more complicated than treating it the same as Pfloatarray here?

File: middle_end/closure/closure.ml (hunk 1 above)
Comment (mshinwell) [reply middle_end/closure/closure.ml]: This was a stale CR, removed.

File: middle_end/flambda/simplify_primitives.ml (hunk 2 above)
Comment (mshinwell) [reply middle_end/flambda/simplify_primitives.ml]: No, this was also a stale CR.  Flambda 1 uses the Clambda primitives in its IR, so in fact it never sees `Pmakefloatblock` here, as `Convert_primitives` has turned it into `Pmakearray Pfloatarray`.

**Review by stedolan — APPROVED** [review]

**Pull request merged**